lib/publiccloud/provider.pm

```diff
@@ -126,6 +121,94 @@ sub parse_ipa_output {
     return $ret;
 }
 
+=head2 create_ssh_key
 
-1;
+Creates an ssh keypair in a given file path by $args{ssh_private_key_file}
+
+=cut
+sub create_ssh_key {
+    my ($self, %args) = @_;
+    if (script_run('test -f ' . $args{ssh_private_key_file}) != 0) {
+        assert_script_run('SSH_DIR=`dirname ' . $args{ssh_private_key_file} . '`; mkdir -p $SSH_DIR');
+        assert_script_run('ssh-keygen -b 2048 -t rsa -q -N "" -f ' . $args{ssh_private_key_file});
+    }
+}
+
+=head2 run_ipa
+
+called by childs within ipa function
+
+=cut
+sub run_ipa {
+    my ($self, %args) = @_;
+    $args{cleanup}              //= 1;
+    $args{ssh_private_key_file} //= '.ssh/id_rsa';
+    $args{tests}                //= '';
+    $args{timeout}              //= 60 * 20;
+    $args{results_dir}          //= 'ipa_results';
+    $args{distro}               //= 'sles';
+    $args{tests} =~ s/,/ /g;
+
+    my $provider = get_var('PUBLIC_CLOUD_PROVIDER');
+
+    my $cmd = 'ipa --no-color test ' . lc $provider;
+    $cmd .= ' --debug ';
+    $cmd .= "--distro " . $args{distro} . " ";
+    $cmd .= '--region "' . $self->region . '" ';
+    $cmd .= '--results-dir "' . $args{results_dir} . '" ';
+    $cmd .= ($args{cleanup}) ? '--cleanup ' : '--no-cleanup ';
+    $cmd .= '--instance-type "' . $args{instance_type} . '" ';
+    $cmd .= '--ssh-private-key-file "' . $args{ssh_private_key_file} . '" ';
+
+    if ($args{credentials_file}) {
```

Comment:
It's quite implicit that azure and google is using a credentials_file so we can call create_ssh_key(), but EC2 need  a SSH_key which was created via aws cli.

Reply:
Not sure I get your comment. This is indeed executed only for azure and google.

Reply:
something like this?
  
    if (! -e $args{ssh_private_key_file}) { 
        $self->create_ssh_key(ssh_private_key_file => $args{ssh_private_key_file});
        $cmd .= '--service-account-file "' . $args{credentials_file} . '" ';
    }

Reply:
I thought just something like this:

``
$self->create_ssh_key(ssh_private_key_file => $args{ssh_private_key_file}) if (! -e $args{ssh_private_key_file});

$cmd .= '--service-account-file "' . $args{credentials_file} . '" ' if ($args{ssh_private_key_file});
``
So if no private key is  given, we create one with the common function.